File: iosApp/iosApp/Modules/KeyStore/KeyStoreWireframeFactory.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

protocol KeyStoreWireframeFactory {
    func make(
        _ parent: UIViewController?
    ) -> KeyStoreWireframe
}

// MARK: - DefaultKeyStoreWireframeFactory

final class DefaultKeyStoreWireframeFactory {

    private let keyStoreService: KeyStoreService
    private let networksService: NetworksService
    private let newMnemonic: MnemonicNewWireframeFactory
    private let updateMnemonic: MnemonicUpdateWireframeFactory
    private let importMnemonic: MnemonicImportWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory

    init(
        keyStoreService: KeyStoreService,
        networksService: NetworksService,
        newMnemonic: MnemonicNewWireframeFactory,
        updateMnemonic: MnemonicUpdateWireframeFactory,
        importMnemonic: MnemonicImportWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory
    ) {
        self.keyStoreService = keyStoreService
        self.networksService = networksService
        self.newMnemonic = newMnemonic
        self.updateMnemonic = updateMnemonic
        self.importMnemonic = importMnemonic
        self.alertWireframeFactory = alertWireframeFactory
    }
}

// MARK: - WalletsWireframeFactory

extension DefaultKeyStoreWireframeFactory: KeyStoreWireframeFactory {

    func make(
        _ parent: UIViewController?
    ) -> KeyStoreWireframe {
        DefaultKeyStoreWireframe(
            parent,
            keyStoreService: keyStoreService,
            networksService: networksService,
            newMnemonic: newMnemonic,
            updateMnemonic: updateMnemonic,
            importMnemonic: importMnemonic,
            alertWireframeFactory: alertWireframeFactory
        )
    }
}

// MARK: - Assembler

final class KeyStoreWireframeFactoryAssembler: AssemblerComponent {

    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> KeyStoreWireframeFactory in
            DefaultKeyStoreWireframeFactory(
                keyStoreService: resolver.resolve(),
                networksService: resolver.resolve(),
                newMnemonic: resolver.resolve(),
                updateMnemonic: resolver.resolve(),
                importMnemonic: resolver.resolve(),
                alertWireframeFactory: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/keystore/KeyStoreWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.keystore

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreService
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3wallet.android.modules.compose.mnemonicimport.MnemonicImportWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.mnemonicnew.MnemonicNewWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.keyStore.KeyStoreWireframe
import smartadapter.internal.extension.name

interface KeyStoreWireframeFactory {
    fun make(parent: Fragment?): KeyStoreWireframe
}

class DefaultKeyStoreWireframeFactory(
    private val keyStoreService: KeyStoreService,
    private val networksService: NetworksService,
    private val mnemonicNewWireframeFactory: MnemonicNewWireframeFactory,
    private val mnemonicImportWireframeFactory: MnemonicImportWireframeFactory,
): KeyStoreWireframeFactory {

    override fun make(parent: Fragment?): KeyStoreWireframe = DefaultKeyStoreWireframe(
        parent?.let { WeakRef(it) },
        keyStoreService,
        networksService,
        mnemonicNewWireframeFactory,
        mnemonicImportWireframeFactory
    )
}

class KeyStoreWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(KeyStoreWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultKeyStoreWireframeFactory(
                it.resolve(KeyStoreService::class.name),
                it.resolve(NetworksService::class.name),
                it.resolve(MnemonicNewWireframeFactory::class.name),
                it.resolve(MnemonicImportWireframeFactory::class.name),
            )
        }
    }
}